File: leetcode/1624. Largest Substring Between Two Equal Characters/1624.cpp
// https://helloacm.com/compute-the-largest-substring-between-two-equal-characters-using-hash-table/
// https://leetcode.com/problems/largest-substring-between-two-equal-characters/
// EASY, HASH MAP

class Solution {
public:
    int maxLengthBetweenEqualCharacters(string s) {
        unordered_map<char, int> data;
        int ans = -1;
        for (int i = 0; i < s.size(); ++ i) {
            if (data.count(s[i])) {
                ans = max(ans, i - data[s[i]] - 1);
            } else {
                data[s[i]] = i;
            }
        }
        return ans;
    }
};

File: leetcode/1624. Largest Substring Between Two Equal Characters/1624.py
# https://helloacm.com/teaching-kids-programming-largest-substring-between-two-equal-characters-hash-map/
# https://leetcode.com/problems/largest-substring-between-two-equal-characters/
# EASY, HASH MAP

class Solution:
    def maxLengthBetweenEqualCharacters(self, s: str) -> int:
        seen = {}
        ans = -1
        for i, v in enumerate(s):
            if v in seen:
                ans = max(ans, i - seen[v] - 1)
            else:
                seen[v] = i
        return ans
